kernel/arch/riscv/cpu_features.c

```diff
@@ -7,13 +7,289 @@
+
+static const char *isa_exts_names[] =
+{
+   "i",
+   "m",
+   "a",
+   "f",
+   "d",
+   "q",
+   "c",
+   "b",
+   "k",
+   "j",
+   "p",
+   "v",
+   "h",
+   "zicbom",
+   "zicboz",
+   "zicntr",
+   "zicsr",
+   "zifencei",
+   "zihintpause",
+   "zihpm",
+   "zba",
+   "zbb",
+   "zbs",
+   "smaia",
+   "ssaia",
+   "sscofpmf",
+   "sstc",
+   "svinval",
+   "svnapot",
+   "svpbmt",
+};
+
+static int fdt_parse_model(void *fdt)
+{
+   int node, len;
+   const fdt32_t *prop;
+
+   node = fdt_path_offset(fdt, "/");
+   if (node < 0)
+      return -1;
+
+   prop = fdt_getprop(fdt, node, "model", &len);
+   if (!prop || !len)
+      return -1;
+
+   strncpy(model_name, (const char *)prop, sizeof(model_name));
+
+   return 0;
+}
+
+static int
+fdt_parse_hart_id(void *fdt, int cpu_offset, u32 *hartid)
+{
+   int len;
+   const void *prop;
+   const fdt32_t *val;
+
+   if (!fdt || cpu_offset < 0)
+      return -1;
+
+   prop = fdt_getprop(fdt, cpu_offset, "device_type", &len);
+   if (!prop || !len)
+      return -1;
+   if (strncmp (prop, "cpu", strlen ("cpu")))
+      return -1;
+
+   val = fdt_getprop(fdt, cpu_offset, "reg", &len);
+   if (!val || len < (int)sizeof(fdt32_t))
+      return -1;
+
+   if (len > (int)sizeof(fdt32_t))
+      val++;
+
+   if (hartid)
+      *hartid = fdt32_to_cpu(*val);
+
+   return 0;
+}
+
+static int fdt_parse_isa_extensions(void *fdt)
+{
+   int cpu_node, cpus_node, len, rc;
+   u32 hart_id;
+   const fdt32_t *prop;
+
+   cpus_node = fdt_path_offset(fdt, "/cpus");
+   if (cpus_node < 0)
+      return -1;
+
+   fdt_for_each_subnode(cpu_node, fdt, cpus_node) {
+
+      rc = fdt_parse_hart_id(fdt, cpu_node, &hart_id);
+      if (rc)
+         return rc;
+
+      prop = fdt_getprop(fdt, cpu_node, "device_type", NULL);
+
+      if (prop
+         && !strcmp((void *)prop, "cpu")
+         && hart_id == get_boothartid()) {
+
+         prop = fdt_getprop(fdt, cpu_node, "riscv,isa", &len);
+         if (prop && len)
+            strncpy(isa_string_buf, (void *)prop,
+                    sizeof(isa_string_buf));
+
+         prop = fdt_getprop(fdt, cpu_node, "riscv,isa-extensions", &len);
+         if (prop && len)
+            memcpy(isa_ext_string_buf, (void *)prop,
+                   MIN((ulong)len, sizeof(isa_ext_string_buf)));
+
+         return 0;
+      }
+   }
+
+   return -1;
+}
+
+static bool isa_ext_string_match(const char *string)
+{
+   char *buf = isa_ext_string_buf;
+   char *end = buf + strnlen(buf, sizeof(isa_ext_string_buf));
+
+   // try isa_ext_string_buf
+
+   for (char *ptr = buf; ptr < end; ptr += strlen(ptr) + 1) {
+
+      if (strcmp(string, ptr) == 0)
+         return true;
+   }
+
+   /*
+    * try isa_string_buf, Some platforms(e.g. qemu-virt) may not have
+    * riscv,isa-extensions property defined in the device tree
+    */
+   buf = isa_string_buf;
+   if (strncmp(buf, "rv64", 4) && strncmp(buf, "rv32", 4))
+      panic("riscv,isa property not defined in device tree\n");
```

Comment:
`panic()` does not require a trailing new line. Also can you replace "riscv," with "riscv: " ?

Reply:
fixed!